normandy/recipes/signing.py

```diff
@@ -190,12 +191,18 @@ def verify_x5u(url, expire_early=None):
     chain. Otherwise, raise an exception explaining why they are not
     valid.
     """
-    req = requests.get(url)
-    req.raise_for_status()
-    pem = req.content.decode()
 
-    der_encoded_certs = extract_certs_from_pem(pem)
-    decoded_certs = [parse_cert_from_der(der) for der in der_encoded_certs]
+    cache_key = f"fetch-x5u-certs::{url}"
+
+    decoded_certs = cache.get(cache_key)
+    if decoded_certs is None:
+        req = requests.get(url, timeout=10)
```

Comment:
Looks like this is called from `recipe_signatures_are_correct`, `action_signatures_are_correct`, and `signatures_use_good_certificates`. Only the latter will actually catch the exception if this times out. The others don't catch any exceptions from requests at all now. This needs fixed so that the heartbeat will still return a response when this fails.

Additionally,  I wonder if you should short-circuit further requests to the same x5u url within a heartbeat request once one has failed. Otherwise in the case of an empty cache during a significant cloudfront outage we risk the heartbeat taking `(signed_recipe + signed_action) * request timeout` , which even if we dropped the timeout from 10 to 1 seconds would cause the same problem we saw yesterday. I don't think there's any benefit to repeating the x5u request for additional recipes or actions, since after the first one fails you want the heartbeat to return a 500, right?

You could probably accomplish this by caching a marker for an error in the request and bailing out of this function if that's present. Ideally this marker would persist only for the life of the hearbeat request, but you could probably accomplish the same effect using django's cache but setting a shorter TTL like 5 seconds.

Reply:
This timeout is really generous. Under normal conditions this will return in under 100ms. I think it could be okay if we short-circuit for failed requests, but otherwise we should make it lower.